docs/authdecode.md

```diff
@@ -0,0 +1,216 @@
+            Check PolyVfyWit(ck, π₂, com_p, (c, pc))
+        // Check q(c)·v(c) == z(c) == p(c) · p¹(c), i.e., that v | z
+            Let vc = v(c)
+            Let 1c = 1ⁿ(c)
+            Check vc · qc == pc · (pc - 1c)
+```
+
+`R_pack` states that the bytestring `P` is the packing of the bitstring `p`, where just `com_P` and `com_p` are given. We follow the generic sigma-protocol framework described in Fig. 3 of "Unifying Zero-Knowledge Proofs of Knowledge" by Maurer. We make a homomorphism φ: 𝔽[X] → (𝔾, 𝔾) such that our plaintext `p` gets mapped to `(com_p, com_P)` where `P = pack(p)`. Thus, proving knowledge of a preimage of `(com_p, com_P)` proves that `P` is the packing of `p`. Note this protocol is very simple and likely very fast, but its overhead is nontrivial: a packing proof of a 4KB (= 32kb) plaintext is itself ~1MB.
+```
+R_pack = {
+    (com_P, com_p, ck; p, P) :
+      PolyVfyOpen(ck, p, com_p)
+    ∧ PolyVfyOpen(P, com_P)
+    ∧ P = pack(p)
+}
+Protocol: // This is almost identical to a Schnorr identity proof
+    Define φ(r) = (
+        rᵢG₁ + ... + r_{8i}·G_{8i}, // Commitment to r
+        q₁G₁ + ... + qₙGₙ,          // Commitment to pack(r)
+    ), where
+         ck = (G₁, ..., G_{8i})
+         qᵢ = r_{8i} + 2·r_{8i+1} + ... + 2⁷·r_{8i+7}
+
+    Prover → Verifier:
+        Sample polyn k uniformly, where deg k = deg p
+        Let K = φ(k)
+        Send K
+    Prover ← Verifier:
+        Sample and send a challenge scalar c
+    Prover → Verifier:
+        Send s = k + c·p
+    Verifier:
+        Check that φ(s) = K + c·(com_p, com_P)
+```
+
+## Note on arithmetic delta
+
+Ordinarily, the output labels of a 2PC garbled circuit computation are 128-bit strings `wᵢ ∈ {0,1}¹²⁸`. But for the protocol above to work, `wᵢ` need to have algebraic structure. Fortunately, we can retrofit this requirement to any garbled-circuit scheme as follows.
+
+Let `Wᵢⱼ' ∈ {0,1}¹²⁸` represent a bitstring label of the underlying garbled-circuit scheme. The garbler constructs its _algebraic labels_ `Wᵢⱼ ∈ 𝔽` as follows. First, it randomly selects a PRG seed `ρ`, and then generates a vector of `n` field elements `W₀ = PRG(ρ)`. The garbler also uniformly selects `Δ ∈ 𝔽`, and sets `W₁ = W₀ + (Δ, Δ, ..., Δ)`. The `W₀` and `W₁` represent the labels of the 0 bits and the 1 bits, respectively. In order to transmit the algebraic labels to the Requester, the Notary encrypts each algebraic label using the nonalgebraic label as a key. Concretely, the Notary computes `Cᵢⱼ = Enc_{Wᵢⱼ'}(Wᵢⱼ)` (using AES-GCM, since `Wᵢⱼ` is two block widths), shuffles `C` such that each `Cᵢ₀` is swapped with `Cᵢ₁` with probability 1/2, and sends `C` to the Requester.
+
+When the underlying 2PC protocol terminates, the Requester will compute `wᵢ = Enc_{wᵢ'}(Cᵢ₀)` and `wᵢ = Enc_{wᵢ'}(Cᵢ₁)`, keeping whichever decryption succeeds.
+
+Thus, we have constructed a post-GC setup whereby each `wᵢ` is a random field element, and each column of `W ∈ 𝔽²ⁿ` is related by a global, uniform Δ.
```

Comment:
It looks like that the garbler could garble an identity gate with input bit-string labels and output arithmetic labels. Then we only need to hash  "encrypt" output labels together with the point-and-permute optimization.

Reply:
Cleaned up in latest version